Week number support

peternowee opened the pull request:
This PR follows on https://github.com/owncloudarchive/calendar/issues/929 and is work in progress.

This branch provides:
- A recent development master of FullCalendar modified with a proposed new feature to display the week numbers in the day cells instead of in a separate column.
- Changes in ownCloud Calendar to accommodate the new feature, such as CSS styling and a user setting to toggle week number display. 

To do:
- [x] Upstream implementation of week number display in day cells: https://github.com/fullcalendar/fullcalendar/pull/3024.
- [x] How to safely update the view after changing the "Show week numbers" setting, i.e. without losing the current date, view, height, events, etc.? See peternowee/calendar@f5c921d commit message. Currently waiting to see what upstream will do with this: https://github.com/fullcalendar/fullcalendar/issues/564, https://github.com/fullcalendar/fullcalendar/issues/2709.
- [ ] Translation of string "Show week numbers".

I will try to keep this branch up-to-date by regularly merging from owncloud/calendar master and fullcalendar/fullcalendar master.

georgehrke commented:
First of all: thanks for your contribution! :)
- [ ] please fix the unit test. extend this with your added parameter
  https://github.com/owncloud/calendar/blob/6287b03694e993247daf5e48eaf8e087ac35312e/tests/unit/controller/viewcontrollerTest.php#L128
- [ ] instead of merging master, please do a `git rebase master` in your feature branch
- [ ] shipping patched 3rdparty libs won't happen, this PR will be put on hold till your week numbers feature made it into an official fullcalendar update. We'll update immediately afterwords and will get your PR in.

> How to safely update the view after changing the "Show week numbers" setting, i.e. without losing the current date, view, height, events, etc.? 

Let's see what upstream does about this, but I'm sure we'll find a solution. Except for the height we can definitely restore everything.

> Translation of string "Show week numbers".

Once this branch was merged, the added strings will appear on transifex and folks over there will translate it.

peternowee commented:
Ok, I will try to fix the unit test and use `git rebase` from now on.

Of course, it was never my intention to ship a patched FullCalendar. I only put it in here to make it easy for others to check out my work before upstream implementation.

Thanks for your response!

peternowee commented:
Just updated this demo:
- Rebased to yesterday's Calendar app git master.
- Using the latest FullCalendar code (2.8.0+git).
- Plus a current snapshot of the week number proposal as it is being discussed with FullCalendar now.

Further changes:
- ownCloud's `setUserValue` did not accept boolean values anymore, so casting to string.
- Adapt `viewcontrollerTest.php` stub method to expect a new method call related to week number setting.

Screenshot:
![20160625 fc 4eef482 pn weeknr-in-daycell d9c2dd8 deutsch german month view](https://cloud.githubusercontent.com/assets/10220583/16357708/60e339ac-3afe-11e6-9989-0dfaa9444058.png)

Additional todo's:
- [x] Reduce font size for week numbers to same size as day numbers.
- [ ] Add unit tests for `getShowWeekNr()` and `setShowWeekNr()` to `settingscontrollerTest.php`.
- [ ] Add Javascript tests.

peternowee commented:
Just updated again:
- Rebased to yesterday's Calendar app git master.
- Using the latest FullCalendar code (2.9.0).
- Plus a current snapshot of the week number proposal as it has now been **accepted** by FullCalendar. (It may still take some time before it is included in a release, because it will be part of a breaking release, see fullcalendar/fullcalendar#3024).

Further changes:
- FullCalendar 2.9.0 now allows dynamic option setting, so the question of how to update the view when the week numbers are toggled on or off is solved!

Screenshot has not changed, so see above.

georgehrke commented:
ownCloud comes with jQuery 2.1.4 and moment.js 2.10.3.
So updating to fullcalendar 3 should not be too difficult

peternowee commented:
Yes, so far it has not been too difficult. I will try to update this branch with the new FullCalendar v3 beta soon.

peternowee commented:
I updated the code once again:
- Rebased to last Tuesday's Calendar app git master.
- Using the latest FullCalendar code (v3-beta), which now supports week numbers in the day cells.

Notes:
- Recently, angular-ui-calendar was removed (owncloud/calendar#680) and the settings service was changed (owncloud/calendar#652). I tried to change my code accordingly. Since I am rebasing here, those changes are not visible in incremental commits. However, most changes were made to commit "Add settings checkbox for showing week numbers". If you checked that one earlier, you might want to check it again now: commit 13416ba.
- Earlier this year, we shortly [discussed the week number formatting for in Day view and Week view](https://github.com/owncloudarchive/calendar/issues/929#issuecomment-176843970). You said you would go with the default (FullCalendar) style. I have now removed the commit with the example formatting, so the default style is now used. Contrary to what I thought at the time, the 'W'-prefix is actually localized by FullCalendar, so that is great.
- To reduce the week number font size, I simply applied the entire CSS block for the day numbers to week numbers as well. However, that block also included `opacity: 0.8`, so the week numbers are a bit lighter right now. Let me know if that is a problem and I will target the week numbers separately, so the additional opacity will not apply to them.

Questions:
- Should the upgrade to FullCalendar 3.0 not be separated from this PR? Who should do that?
- @jancborchardt: Perhaps you want to have another look at the screenshots as well?

Latest screenshots (Month view, Week view, Day view):

![20160819 oc ab29515e fc 20e805b month view deutsch german](https://cloud.githubusercontent.com/assets/10220583/17814555/c0273e7e-6630-11e6-934c-db4a60fa9b00.png)
![20160819 oc ab29515e fc 20e805b week view deutsch german](https://cloud.githubusercontent.com/assets/10220583/17814559/c27b1420-6630-11e6-8209-1314f4f45461.png)
![20160819 oc ab29515e fc 20e805b day view deutsch german](https://cloud.githubusercontent.com/assets/10220583/17814565/c823b4fe-6630-11e6-9a16-ef6d8d351d52.png)

georgehrke commented:
Update to FullCalendar 3 is in https://github.com/owncloud/calendar/pull/734 :)

peternowee committed:
Rename lang to locale for FullCalendar 3.0.0

In FullCalendar 3.0.0, the `lang` option was renamed to `locale`.

peternowee committed:
Turn on week numbers

peternowee committed:
Reduce font size for week numbers as well

peternowee committed:
Add settings checkbox for showing week numbers

This adds a checkbox under Settings to "Show week numbers". The setting
is saved as Calendar-app user setting `currentShowWeekNr`.

peternowee committed:
Adapt stub method getUserValue to week number setting

This commit adapts the stub method for `getUserValue()` in the PHPUnit
test for viewcontroller to expect another call to retrieve the
`currentShowWeekNr` setting.

peternowee commented:
I rebased my code again to current master:
- Rebased to last Wednesday's Calendar app git master.
- Since the update to FullCalendar 3.0 is in the Calendar app master branch now (see Georg's comment above), there is no commit for that in this branch anymore.

Notes:
- I noticed another checkbox option was added to master (`skipPopover`, commit c4ee5b33), which was coded a bit differently than my original checkbox option for showing week numbers. I changed my code to be similar to that for `skipPopover`. Again, since this is a rebase, those changes are not visible in incremental commits. However, most were in commit "Add settings checkbox for showing week numbers", so if you checked that one earlier, you might want to check it again now: commit 8cfba37b.
- Screenshots did not change, so see above.
- @georgehrke I still intend to add some more unit tests, but I have limited time. I understand you are trying to get this into 1.4? Around when would you need those tests?

georgehrke commented:
closing in favor of https://github.com/owncloud/calendar/pull/737 :)
